Fix propagating of unhandled exception to native caller

janvorli opened the pull request:
The final propagation of unhandled exception into topmost native frames on foreign threads has a bug that hits when the native caller catches the exception and then calls into managed code again. The problem is that we weren't popping out the explicit frames and ExInfos. So when the managed code is entered again, the `Thread::m_pFrame` points to a garbage. The same is true for the `ExInfo` chain.

This is causing crashes in a 3rd party application where a foreign thread runs some MFC code that calls to some managed functions regularly and swallows exceptions stemming from it.

While investigating the issue, I have found that there is a secondary related problem. When we mark the thread by setting the TSNC_ProcessedUnhandledException flag, the ProcessCLRException intentionally starts ignoring managed frames, because we try to let the exception bubble out to the 3rd party code. But if it catches the exception and calls back to managed code, the
TSNC_ProcessedUnhandledException flag stays set and thus when the exception would need to propagate through native frames, all the managed frames would be ignored.

This change fixes both of the problems and adds a test that crashes without the fix in both of the cases.

Close #113751

janvorli committed:
Fix propagating of unhandled exception to native caller

The final propagation of unhandled exception into topmost native frames
on foreign threads has a bug that hits when the native caller catches
the exception and then calls into managed code again.
The problem is that we weren't popping out the explicit frames and
ExInfos. So when the managed code is entered again, the `Thread::m_pFrame`
points to a garbage. The same is true for the `ExInfo` chain.

This is causing crashes in a 3rd party application where a foreign
thread runs some MFC code that calls to some managed functions regularly
and swallows exceptions stemming from it.

While investigating the issue, I have found that there is a secondary
related problem. When we mark the thread by setting the
TSNC_ProcessedUnhandledException flag, the ProcessCLRException
intentionally starts ignoring managed frames, because we try to let the
exception bubble out to the 3rd party code. But if it catches the
exception and calls back to managed code, the
TSNC_ProcessedUnhandledException flag stays set and thus when the
exception would need to propagate through native frames, all the managed
frames would be ignored.

This change fixes both of the problems and adds a test that crashes
without the fix in both of the cases.

Close #113751

jkotas commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -934,11 +937,30 @@ ProcessCLRExceptionNew(IN     PEXCEPTION_RECORD   pExceptionRecord,
+    {
+        // We are in the 1st pass of exception handling, but the thread mark says that it has already executed 2nd pass
+        // of unhandled exception handling. That means that some external native code on top of the stack has caught the
+        // exception that runtime considered to be unhandledand a new native exception was thrown on the current thread.
```
```suggestion
        // exception that runtime considered to be unhandled, and a new native exception was thrown on the current thread.
```

janvorli committed:
Make the test Windows only and fix comment spelling

janvorli committed:
Fix Windows build of the test

janvorli committed:
Disable the added testing for Mono

janvorli commented:
I can see the change has broken two diagnostic tests. I am looking into it

janvorli committed:
Move the frames popping to 2nd pass

The SOS tests that examine crash dump of an unhandled exception were
failing with this change due to the fact that the explicit frames and
ExInfos were popped before the RaiseException. The Watson / crash dump
generation kicks in at the end of the first pass of the RaiseException,
so the frames need to be popped in the 2nd pass. I have changed it so
that they are popped at the first call to the ProcessCLRException on the
2nd pass.

janvorli commented:
@jkotas SOS tests ran in the CI were failing because the frames were popped too early. They need to stay until the RaiseException completes the first pass so that a crash dump on Windows contains them.

janvorli commented:
/ba-g the test failures are infra issues on azurelinux that occur on all recent CI runs.
